### PR TITLE
fix "silent" option for synchronous replication

### PR DESCRIPTION
**Pull request by jsteemann**
the previous change of mine did not work properly

### CODE REVIEW

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/129/

File: arangod/Transaction/Methods.cpp [hunk 1]
```diff
@@ -1712,121 +1721,120 @@ OperationResult transaction::Methods::insertLocal(
...
+
+    // Now replicate the good operations on all followers:
+    std::string path =
+      "/_db/" + arangodb::basics::StringUtils::urlEncode(vocbase().name()) +
+      "/_api/document/" + arangodb::basics::StringUtils::urlEncode(collection->name()) +
+      "?isRestore=true&isSynchronousReplication=" + ServerState::instance()->getId() +
+      "&" + StaticStrings::SilentString + "=true" + 
+      "&" + StaticStrings::OverWrite + "=" + (options.overwrite ? "true" : "false");
+
+    VPackBuilder payload;
```
Comment (graetzer) [inline arangod/Transaction/Methods.cpp]: use a BuilderLeaser ??

File: arangod/Transaction/Methods.cpp [hunk 2]
```diff
@@ -2079,127 +2091,126 @@ OperationResult transaction::Methods::modifyLocal(
...
+        "/_api/document/" +
+        arangodb::basics::StringUtils::urlEncode(collection->name()) +
+        "?isRestore=true&isSynchronousReplication=" +
+        ServerState::instance()->getId() + "&" + StaticStrings::SilentString + "=true";
 
-    if (doingSynchronousReplication) {
-      // In the multi babies case res is always TRI_ERROR_NO_ERROR if we
-      // get here, in the single document case, we do not try to replicate
-      // in case of an error.
+      VPackBuilder payload;
```
Comment (graetzer) [inline arangod/Transaction/Methods.cpp]: use BuilderLeaser

File: arangod/Transaction/Methods.cpp [hunk 3]
```diff
@@ -2369,125 +2387,124 @@ OperationResult transaction::Methods::removeLocal(
...
-          VPackArrayIterator itResult(ourResult);
-          while (itValue.valid() && itResult.valid()) {
-            TRI_ASSERT((*itResult).isObject());
-            if (!(*itResult).hasKey(StaticStrings::Error)) {
-              doOneDoc(itValue.value(), itResult.value());
-              count++;
-            }
-            itValue.next();
-            itResult.next();
+      VPackBuilder payload;
```
Comment (graetzer) [inline arangod/Transaction/Methods.cpp]: use BuilderLeaser

File: arangod/Transaction/Methods.cpp [hunk 4]
```diff
@@ -2640,69 +2667,63 @@ OperationResult transaction::Methods::truncateLocal(
...
-        // nullptr only happens on controlled shutdown
-        std::string path =
-            "/_db/" +
-            arangodb::basics::StringUtils::urlEncode(vocbase().name()) +
-            "/_api/collection/" +
-            arangodb::basics::StringUtils::urlEncode(collectionName) +
-            "/truncate?isSynchronousReplication=" +
-            ServerState::instance()->getId();
-        auto body = std::make_shared<std::string>();
+    if (cc != nullptr) {
```
Comment (graetzer) [inline arangod/Transaction/Methods.cpp]: isn't this just happening in the case of a shutdown ?

File: arangod/Transaction/Methods.h [hunk 5]
```diff
@@ -128,8 +128,6 @@ class Methods {
   using VPackBuilder = arangodb::velocypack::Builder;
   using VPackSlice = arangodb::velocypack::Slice;
 
-  double const TRX_FOLLOWER_TIMEOUT = 3.0;
```
Comment (graetzer) [inline arangod/Transaction/Methods.h]: 👍

File: arangod/Transaction/Methods.cpp [hunk 6]
```diff
@@ -2369,125 +2387,124 @@ OperationResult transaction::Methods::removeLocal(
...
           }
         }
       }
     }
   }
+  
+  // wait for operation(s) to be synced to disk here. On rocksdb maxTick == 0
+  if (res.ok() && options.waitForSync && maxTick > 0 &&
+      isSingleOperationTransaction()) {
+    EngineSelectorFeature::ENGINE->waitForSyncTick(maxTick);
```
Comment (graetzer) [inline arangod/Transaction/Methods.cpp]: isn't this a noop on rocksdb ?

Comment (jsteemann) [reply arangod/Transaction/Methods.cpp]: Yes for RocksDB, but not for MMFiles.

File: arangod/Transaction/Methods.cpp (hunk 4 above)
Comment (jsteemann) [reply arangod/Transaction/Methods.cpp]: yes. But we must not call something on the nullptr, so we need the check.

Comment (graetzer) [reply arangod/Transaction/Methods.cpp]: why not ```if (cc == nullptr) return TRI_ERROR_SHUTTING_DOWN```

Comment (jsteemann) [reply arangod/Transaction/Methods.cpp]: there is some code running after the code, so we cannot retun from here

**Pull request merged**